[Merged by Bors] - feat: integrability of log ‖meromorphic‖

kebekus opened the pull request:
For real-meromorphic functions `f`, establish interval integrability of `log ‖f ·‖`. In particular, show that `log ∘ sin` is interval integrable. For complex-meromorphic functions `f`, establish circle integrability of `log ‖f ·‖`.

At the suggestion of @sgouezel, split the file [Mathlib/Analysis/SpecialFunctions/Integrals.lean](https://github.com/leanprover-community/mathlib4/pull/24845/files#diff-866dd3a7c4603a1a9d9361ee9e58ea2a97ab83d2037e34d5a8ac23862f421c65). The file establishes values of special interval integrals. As a prerequisite, it establishes interval integrability of numerous functions. Interval integrability of `log ‖f ·‖`, which is established in the current PR, naturally fits in there. Adding the material to [Mathlib/Analysis/SpecialFunctions/Integrals.lean](https://github.com/leanprover-community/mathlib4/pull/24845/files#diff-866dd3a7c4603a1a9d9361ee9e58ea2a97ab83d2037e34d5a8ac23862f421c65) did however not seem advisable:

- The line of argumentation is different from what is used already, leading to massive new imports and cross-dependencies.
- The current file mixes two lines of results of independent interest that have only a clear one-way dependency (integral computation uses integrability)
- The file was very long already.

As a clean solution, move the material on integrability to a new directory, `Mathlib/Analysis/SpecialFunctions/Integrability`, which contains a file with the original material and another one with the new material. That way, the results on integrability are all in one place and cross-imports are avoided.

This material is used in [Project VD](https://github.com/kebekus/ProjectVD), which aims to formalize Value Distribution Theory for meromorphic functions on the complex plane. Integrability of `log ∘ sin` might be useful in future discussions of polylogarithms.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

sgouezel commented on Mathlib/Analysis/SpecialFunctions/Complex/Analytic.lean:
```diff
@@ -152,4 +153,20 @@ lemma AnalyticOn.log (fs : AnalyticOn ℝ f s) (m : ∀ x ∈ s, 0 < f x) :
     AnalyticOn ℝ (fun z ↦ Real.log (f z)) s :=
   fun z n ↦ (analyticAt_log (m z n)).analyticWithinAt.comp (fs z n) m
 
+/-- The function `Real.sin` is real analytic. -/
+@[fun_prop]
+lemma analyticAt_sin : AnalyticAt ℝ Real.sin x := by
+  have : Real.sin = fun x : ℝ ↦ (Complex.sin x).re := by
+    ext x
+    rfl
+  rw [this]
+  apply AnalyticAt.re_ofReal
+  apply Complex.analyticOnNhd_univ_iff_differentiable.mpr Complex.differentiable_sin
+  trivial
+
+/-- The function `Real.sin` is real analytic. -/
+theorem analyticOnNhd_sin : AnalyticOnNhd ℝ Real.sin Set.univ := by
+  intro x
+  simp [analyticAt_sin]
+
```
```suggestion

/-- The function `Real.sin` is real analytic. -/
@[fun_prop]
lemma analyticAt_sin : AnalyticAt ℝ Real.sin x :=
  Real.contDiff_sin.contDiffAt.analyticAt

/-- The function `Real.sin` is real analytic. -/
theorem analyticOnNhd_sin : AnalyticOnNhd ℝ Real.sin s :=
  fun _ _ ↦ analyticAt_sin
```
We already have this fact (in the form that sin is C^n for any n, including n = omega), so better use the already existing fact. Also, I've stated it for any set instead of just `univ`, because often in applications you need a smaller set.

Could you also register the `AnalyticWithinAt` and `AnalyticOn` versions? Also for `cos`, and `cosh`, and `sinh`, and `exp`, real and complex versions? (I know it's a lot to ask, but that's the way to construct comprehensive API and make sure nobody will need to come back to it later on).
Also, this would better fit in the file `Analysis.SpecialFunctions.Trigonometric.Deriv`.

kebekus replied:
Will do, but probably not before Monday.

kebekus replied:
Implemented.

sgouezel commented on Mathlib/Analysis/SpecialFunctions/Integrals.lean:
```diff
@@ -3,7 +3,10 @@ Copyright (c) 2021 Benjamin Davidson. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Benjamin Davidson
 -/
+import Mathlib.Analysis.Meromorphic.FactorizedRational
```
When you need to add a lot of imports to a file, it's a sign that in fact you should instead create a new file.

kebekus replied:
Implemented.

sgouezel commented on Mathlib/Analysis/SpecialFunctions/Integrals.lean:
```diff
@@ -258,6 +261,68 @@ theorem intervalIntegrable_log' : IntervalIntegrable log volume a b := by
     simpa
 
+/--
+If `f` is meromorphic over `ℝ`, then functions of the form `log ‖f ·‖` are
+interval integrable.
```
```suggestion
If `f` is meromorphic on a compact interval, then functions of the form `log ‖f ·‖` are
interval integrable on this interval.
```

sgouezel commented on Mathlib/Analysis/SpecialFunctions/Integrals.lean:
```diff
@@ -266,6 +331,17 @@ theorem intervalIntegrable_sin : IntervalIntegrable sin μ a b :=
   continuous_cos.intervalIntegrable a b
 
+/--
+The function `log ∘ sin` is invervalintegrable over every interval.
```
```suggestion
The function `log ∘ sin` is interval integrable over every interval.
```

sgouezel commented on Mathlib/MeasureTheory/Integral/CircleIntegral.lean:
```diff
@@ -183,6 +183,22 @@ nonrec theorem add (hf : CircleIntegrable f c R) (hg : CircleIntegrable g c R) :
   hf.add hg
 
+/-- Sums of circle integrable functions are circle integrable. -/
+theorem sum {ι : Type*} (s : Finset ι) {f : ι → ℂ → E} (h : ∀ i ∈ s, CircleIntegrable (f i) c R) :
```
```suggestion
protected theorem sum {ι : Type*} (s : Finset ι) {f : ι → ℂ → E} (h : ∀ i ∈ s, CircleIntegrable (f i) c R) :
```

sgouezel commented on Mathlib/MeasureTheory/Integral/CircleIntegral.lean:
```diff
@@ -183,6 +183,22 @@ nonrec theorem add (hf : CircleIntegrable f c R) (hg : CircleIntegrable g c R) :
+  exact IntervalIntegrable.sum s h
+
+/-- Finsums of circle integrable functions are circle integrable. -/
+theorem finsum {ι : Type*} {f : ι → ℂ → E} (h : ∀ i, CircleIntegrable (f i) c R) :
```
```suggestion
protected theorem finsum {ι : Type*} {f : ι → ℂ → E} (h : ∀ i, CircleIntegrable (f i) c R) :
```

sgouezel commented on Mathlib/MeasureTheory/Integral/IntervalIntegral/Basic.lean:
```diff
@@ -265,6 +265,16 @@ theorem sum (s : Finset ι) {f : ι → ℝ → E} (h : ∀ i ∈ s, IntervalInt
 
+/-- Finsums of interval integrable functions are interval integrable. -/
+@[simp]
+theorem finsum {f : ι → ℝ → E} (h : ∀ i, IntervalIntegrable (f i) μ a b) :
```
```suggestion
protected theorem finsum {f : ι → ℝ → E} (h : ∀ i, IntervalIntegrable (f i) μ a b) :
```

kebekus commented on Mathlib/Analysis/Complex/Meromorphic/CircleIntegrability.lean:
```diff
@@ -0,0 +1,76 @@
+open Filter MeromorphicOn Metric Real
+
+/--
+If `f` is meromorphic on the complex plane, then `log ‖f ·‖` is circle integrable over every circle.
```
Oops. That has been corrected.

kebekus commented on Mathlib/Analysis/SpecialFunctions/Trigonometric/Deriv.lean:
```diff
@@ -174,6 +174,23 @@ theorem deriv_ccos (hc : DifferentiableAt ℂ f x) :
 
+/-- The function `Complex.cos` is complex analytic. -/
+@[fun_prop]
+lemma analyticAt_ccos : AnalyticAt ℂ Complex.cos x :=
```
Done.

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

kebekus commented on Mathlib/Analysis/SpecialFunctions/Trigonometric/Deriv.lean:
```diff
@@ -53,6 +53,23 @@ theorem differentiable_sin : Differentiable ℂ sin := fun x => (hasDerivAt_sin
 
+/-- The function `Complex.sin` is complex analytic. -/
+@[fun_prop]
+lemma analyticAt_sin {x : ℂ} : AnalyticAt ℂ Complex.sin x :=
```
Implemented.

sgouezel commented on Mathlib/Analysis/SpecialFunctions/Integrability/Basic.lean:
```diff
@@ -0,0 +1,283 @@
+import Mathlib.Analysis.SpecialFunctions.Pow.Deriv
+
+/-!
+# Integration of specific interval integrals
```
This docstring is the docstring of the other, old, file. It does not correspond to the contents of the current file. Could you replace it with something more appropriate?

kebekus replied:
Done.

kebekus commented:
@sgouezel I implemented your suggestion and removed the "awaiting author" label.

At present, the "Post PR summary comment" fails with an error message
```
./scripts/update_PR_comment.sh: line 36: GITHUB_TOKEN: unbound variable
```
I'm assuming this is an issue in the build system unrelated to my PR, so I will continue as normal.

grunweg commented:
The failing job is unrelated to your PR: can you merge master, please? Then it should disappear and your PR will be visible on the review queue again. (Sorry for the inconvenience; this is caused by a recent changes to mathlib's continuous integration. An in-progress PR will make such situations impossible in the future.)

sgouezel commented on Mathlib/Analysis/SpecialFunctions/Integrability/LogMeromorphic.lean:
```diff
@@ -0,0 +1,172 @@
+    apply CircleIntegrable.add
+    · apply CircleIntegrable.finsum
+      intro i
+      unfold CircleIntegrable
```
```suggestion
```
`unfold` is not needed if you use `apply` afterwards, as `apply` sees through definitions.

sgouezel reviewed (commented):
Looks good to me, thanks!

kebekus commented on Mathlib/Analysis/SpecialFunctions/Integrability/LogMeromorphic.lean:
```diff
@@ -0,0 +1,172 @@
+If `f` is real-meromorphic on a compact interval, then `log ‖f ·‖` is interval integrable on this
+interval.
+-/
+@[simp]
```
Removed the `simp`s, as suggested.

Ruben-VandeVelde commented:
There's a bunch of code movement in this PR; could you add a note to the PR summary to explain why? (or move these changes to a separate PR)

kebekus commented:
> There's a bunch of code movement in this PR; could you add a note to the PR summary to explain why? (or move these changes to a separate PR)

@Ruben-VandeVelde Thanks for looking at this PR. I updated the description as follows.

At the suggestion of @sgouezel, split the file [Mathlib/Analysis/SpecialFunctions/Integrals.lean](https://github.com/leanprover-community/mathlib4/pull/24845/files#diff-866dd3a7c4603a1a9d9361ee9e58ea2a97ab83d2037e34d5a8ac23862f421c65). The file establishes values of special interval integrals. As a prerequisite, it establishes interval integrability of numerous functions. Interval integrability of `log ‖f ·‖`, which is established in the current PR, naturally fits in there. Adding the material to [Mathlib/Analysis/SpecialFunctions/Integrals.lean](https://github.com/leanprover-community/mathlib4/pull/24845/files#diff-866dd3a7c4603a1a9d9361ee9e58ea2a97ab83d2037e34d5a8ac23862f421c65) did however not seem advisable:

- The line of argumentation is different from what is used already, leading to massive new imports and cross-dependencies.
- The current file mixes two lines of results of independent interest that have only a clear one-way dependency (integral computation uses integrability)
- The file was very long already.

As a clean solution, move the material on integrability to a new directory, `Mathlib/Analysis/SpecialFunctions/Integrability`, which contains a file with the original material and another one with the new material. That way, the results on integrability are all in one place and cross-imports are avoided.

sgouezel commented on Mathlib/Analysis/SpecialFunctions/Integrability/LogMeromorphic.lean:
```diff
@@ -0,0 +1,161 @@
+## Interval Integrability for Logarithms of Real Meromorphic Functions
+-/
+
+namespace intervalIntegral
```
```suggestion
section IntervalIntegrable
```
No need for a namespace, as all the names already mention that you're talking about interval integrability.

sgouezel commented on Mathlib/Analysis/SpecialFunctions/Integrability/LogMeromorphic.lean:
```diff
@@ -0,0 +1,161 @@
+## Circle Integrability for Logarithms of Complex Meromorphic Functions
+-/
+
+namespace circleIntegral
```
```suggestion
section CircleIntegrable
```

sgouezel commented on Mathlib/Analysis/SpecialFunctions/Integrability/LogMeromorphic.lean:
```diff
@@ -0,0 +1,161 @@
+If `f` is complex meromorphic on a circle in the complex plane, then `log ‖f ·‖` is circle
+integrable over that circle.
+-/
+theorem circleIntegrable_log_norm_meromorphicOn (hf : MeromorphicOn f (sphere c |R|)) :
```
I would guess that in 95% of applications, R will be nonnegative, and the assumption will be of the form `MeromorphicOn f (sphere c R)`. Could you also give the statement in this form, called `circleIntegrable_log_norm_meromorphicOn_of_nonneg`?

kebekus replied:
Done.

sgouezel commented on Mathlib/Analysis/SpecialFunctions/Integrability/LogMeromorphic.lean:
```diff
@@ -0,0 +1,161 @@
+If `f` is complex meromorphic on a circle in the complex plane, then `log⁺ ‖f ·‖` is circle
+integrable over that circle.
+-/
+theorem circleIntegrable_posLog_norm_meromorphicOn (hf : MeromorphicOn f (sphere c |R|)) :
```
ditto

kebekus replied:
Done.

kebekus commented on Mathlib/Analysis/SpecialFunctions/Integrability/LogMeromorphic.lean:
```diff
@@ -0,0 +1,161 @@
+Special case of `MeromorphicOn.intervalIntegrable_log`: The function `log ∘ sin` is interval
+integrable over every interval.
+-/
+theorem intervalIntegrable_log_sin : IntervalIntegrable (log ∘ sin) volume a b :=
```
Done

kebekus commented:
@sgouezel @j-loreaux How shall we continue? If you prefer, I can easily split this PR into several pieces. Just let me know if this is what you want me to do.
